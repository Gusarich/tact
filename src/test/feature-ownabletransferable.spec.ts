import { Address, toNano } from '@ton/core';
import { ContractSystem } from '@tact-lang/emulator';
import { __DANGER_resetNodeId } from '../grammar/ast';
import { OwnableTransferableTest } from './features/output/ownabletransferable_OwnableTransferableTest';

describe('feature-ordering', () => {
    beforeEach(() => {
        __DANGER_resetNodeId();
    });
    it('should deploy correctly', async () => {
        // Init
        const system = await ContractSystem.create();
        const treasure = system.treasure('treasure');
        const contract = system.open(
            await OwnableTransferableTest.fromInit(treasure.address)
        );
        const tracker = system.track(contract.address);
        await contract.send(treasure, { value: toNano('1') }, 'test');
        await system.run();
        expect(tracker.collect()).toMatchSnapshot();
        expect(
            (await contract.getOwner()).equals(treasure.address)
        ).toBeTruthy();
    });

    it('should transfer ownership correctly', async () => {
        // Init
        const system = await ContractSystem.create();
        const treasure = system.treasure('treasure');
        const contract = system.open(
            await OwnableTransferableTest.fromInit(treasure.address)
        );
        const tracker = system.track(contract.address);
        await contract.send(treasure, { value: toNano('1') }, 'test');
        await system.run();
        const anotherTreasure = system.treasure('anotherTreasure');
        await contract.send(
            treasure,
            { value: toNano('1') },
            {
                $$type: 'ChangeOwner',
                newOwner: anotherTreasure.address,
                queryId: 0n,
            }
        );
        await system.run();
        expect(tracker.collect()).toMatchSnapshot();
        expect(
            (await contract.getOwner()).equals(anotherTreasure.address)
        ).toBeTruthy();
    });

    it('should renounce ownership correctly', async () => {
        // Init
        const system = await ContractSystem.create();
        const treasure = system.treasure('treasure');
        const contract = system.open(
            await OwnableTransferableTest.fromInit(treasure.address)
        );
        const tracker = system.track(contract.address);
        await contract.send(treasure, { value: toNano('1') }, 'test');
        await system.run();

        const anotherTreasure = system.treasure('anotherTreasure');
        await contract.send(
            treasure,
            { value: toNano('1') },
            {
                $$type: 'ChangeOwner',
                newOwner: anotherTreasure.address,
                queryId: 0n,
            }
        );
        await system.run();
        expect(
            (await contract.getOwner()).equals(anotherTreasure.address)
        ).toBeTruthy();

        await contract.send(
            anotherTreasure,
            { value: toNano('1') },
            {
                $$type: 'RenounceOwnership',
                queryId: 0n,
            }
        );
        await system.run();
        expect(
            (await contract.getOwner()).equals(
                Address.parseRaw(
                    '0:0000000000000000000000000000000000000000000000000000000000000000'
                )
            )
        ).toBeTruthy();
        expect(tracker.collect()).toMatchSnapshot();
    });
});
