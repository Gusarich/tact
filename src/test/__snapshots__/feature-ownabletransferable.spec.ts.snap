// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`feature-ordering should deploy correctly 1`] = `
[
  {
    "$seq": 0,
    "events": [
      {
        "$type": "deploy",
      },
      {
        "$type": "received",
        "message": {
          "body": {
            "text": "test",
            "type": "text",
          },
          "bounce": true,
          "from": "@treasure(treasure)",
          "to": "kQBGF0CUxwjQr65xgop8Bw8l5v1BfLn6Oey9WClezcq-HQ9z",
          "type": "internal",
          "value": "1",
        },
      },
      {
        "$type": "processed",
        "gasUsed": 4870n,
      },
    ],
  },
]
`;

exports[`feature-ordering should renounce ownership correctly 1`] = `
[
  {
    "$seq": 0,
    "events": [
      {
        "$type": "deploy",
      },
      {
        "$type": "received",
        "message": {
          "body": {
            "text": "test",
            "type": "text",
          },
          "bounce": true,
          "from": "@treasure(treasure)",
          "to": "kQBGF0CUxwjQr65xgop8Bw8l5v1BfLn6Oey9WClezcq-HQ9z",
          "type": "internal",
          "value": "1",
        },
      },
      {
        "$type": "processed",
        "gasUsed": 4870n,
      },
    ],
  },
  {
    "$seq": 1,
    "events": [
      {
        "$type": "storage-charged",
        "amount": "0.000000006",
      },
      {
        "$type": "received",
        "message": {
          "body": {
            "type": "known",
            "value": {
              "$$type": "ChangeOwner",
              "newOwner": kQAq4dsn8XD1ot3Hg_-nHdRkM5P1PaOZAy-YLG4p3mkQ96xD,
              "queryId": 0n,
            },
          },
          "bounce": true,
          "from": "@treasure(treasure)",
          "to": "kQBGF0CUxwjQr65xgop8Bw8l5v1BfLn6Oey9WClezcq-HQ9z",
          "type": "internal",
          "value": "1",
        },
      },
      {
        "$type": "processed",
        "gasUsed": 8799n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "type": "known",
              "value": {
                "$$type": "ChangeOwnerOk",
                "newOwner": kQAq4dsn8XD1ot3Hg_-nHdRkM5P1PaOZAy-YLG4p3mkQ96xD,
                "queryId": 0n,
              },
            },
            "bounce": true,
            "from": "kQBGF0CUxwjQr65xgop8Bw8l5v1BfLn6Oey9WClezcq-HQ9z",
            "to": "@treasure(treasure)",
            "type": "internal",
            "value": "0.989738",
          },
        ],
      },
    ],
  },
  {
    "$seq": 2,
    "events": [
      {
        "$type": "storage-charged",
        "amount": "0.000000006",
      },
      {
        "$type": "received",
        "message": {
          "body": {
            "type": "known",
            "value": {
              "$$type": "RenounceOwnership",
              "queryId": 0n,
            },
          },
          "bounce": true,
          "from": "@treasure(anotherTreasure)",
          "to": "kQBGF0CUxwjQr65xgop8Bw8l5v1BfLn6Oey9WClezcq-HQ9z",
          "type": "internal",
          "value": "1",
        },
      },
      {
        "$type": "processed",
        "gasUsed": 11048n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "type": "known",
              "value": {
                "$$type": "ChangeOwnerOk",
                "newOwner": kQAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAHTW,
                "queryId": 0n,
              },
            },
            "bounce": true,
            "from": "kQBGF0CUxwjQr65xgop8Bw8l5v1BfLn6Oey9WClezcq-HQ9z",
            "to": "@treasure(anotherTreasure)",
            "type": "internal",
            "value": "0.987489",
          },
        ],
      },
    ],
  },
]
`;

exports[`feature-ordering should transfer ownership correctly 1`] = `
[
  {
    "$seq": 0,
    "events": [
      {
        "$type": "deploy",
      },
      {
        "$type": "received",
        "message": {
          "body": {
            "text": "test",
            "type": "text",
          },
          "bounce": true,
          "from": "@treasure(treasure)",
          "to": "kQBGF0CUxwjQr65xgop8Bw8l5v1BfLn6Oey9WClezcq-HQ9z",
          "type": "internal",
          "value": "1",
        },
      },
      {
        "$type": "processed",
        "gasUsed": 4870n,
      },
    ],
  },
  {
    "$seq": 1,
    "events": [
      {
        "$type": "storage-charged",
        "amount": "0.000000006",
      },
      {
        "$type": "received",
        "message": {
          "body": {
            "type": "known",
            "value": {
              "$$type": "ChangeOwner",
              "newOwner": kQAq4dsn8XD1ot3Hg_-nHdRkM5P1PaOZAy-YLG4p3mkQ96xD,
              "queryId": 0n,
            },
          },
          "bounce": true,
          "from": "@treasure(treasure)",
          "to": "kQBGF0CUxwjQr65xgop8Bw8l5v1BfLn6Oey9WClezcq-HQ9z",
          "type": "internal",
          "value": "1",
        },
      },
      {
        "$type": "processed",
        "gasUsed": 8799n,
      },
      {
        "$type": "sent",
        "messages": [
          {
            "body": {
              "type": "known",
              "value": {
                "$$type": "ChangeOwnerOk",
                "newOwner": kQAq4dsn8XD1ot3Hg_-nHdRkM5P1PaOZAy-YLG4p3mkQ96xD,
                "queryId": 0n,
              },
            },
            "bounce": true,
            "from": "kQBGF0CUxwjQr65xgop8Bw8l5v1BfLn6Oey9WClezcq-HQ9z",
            "to": "@treasure(treasure)",
            "type": "internal",
            "value": "0.989738",
          },
        ],
      },
    ],
  },
]
`;
